src/Sema.zig

```diff
@@ -22462,20 +22462,18 @@ fn zirErrorCast(sema: *Sema, block: *Block, extended: Zir.Inst.Extended.InstData
     const base_operand_ty = sema.typeOf(operand);
     const dest_tag = base_dest_ty.zigTypeTag(mod);
     const operand_tag = base_operand_ty.zigTypeTag(mod);
-    if (dest_tag != operand_tag) {
-        return sema.fail(block, src, "expected source and destination types to match, found '{s}' and '{s}'", .{
-            @tagName(operand_tag), @tagName(dest_tag),
-        });
-    } else if (dest_tag != .ErrorSet and dest_tag != .ErrorUnion) {
+
+    if (dest_tag != .ErrorSet and dest_tag != .ErrorUnion) {
         return sema.fail(block, src, "expected error set or error union type, found '{s}'", .{@tagName(dest_tag)});
     }
-    const dest_ty, const operand_ty = if (dest_tag == .ErrorUnion) .{
-        base_dest_ty.errorUnionSet(mod),
-        base_operand_ty.errorUnionSet(mod),
-    } else .{
-        base_dest_ty,
-        base_operand_ty,
-    };
+    if (operand_tag != .ErrorSet and operand_tag != .ErrorUnion) {
+        return sema.fail(block, src, "expected error set or error union type, found '{s}'", .{@tagName(operand_tag)});
+    }
+    if (dest_tag == .ErrorSet and operand_tag == .ErrorUnion) {
+        return sema.fail(block, src, "cannot cast an error union type to error set", .{});
+    }
```

Comment:
Could you make sure these have compile error test coverage? See the files in `test/cases/compile_errors/`.